ES: useMultitypeIndex is not set to true on migrated JanusGraph instance

Testing out a migration from JanusGraph 0.1.1 to JanusGraph 0.2.0 with Elasticsearch 1.5.1, ran into this when opening up the graph:

```
java.lang.IllegalArgumentException: Could not instantiate implementation: org.janusgraph.diskstorage.es.ElasticSearchIndex
	at org.janusgraph.util.system.ConfigurationUtil.instantiate(ConfigurationUtil.java:69)
	at org.janusgraph.diskstorage.Backend.getImplementationClass(Backend.java:477)
	at org.janusgraph.diskstorage.Backend.getIndexes(Backend.java:464)
	at org.janusgraph.diskstorage.Backend.<init>(Backend.java:149)
	at org.janusgraph.graphdb.configuration.GraphDatabaseConfiguration.getBackend(GraphDatabaseConfiguration.java:1897)
	at org.janusgraph.graphdb.database.StandardJanusGraph.<init>(StandardJanusGraph.java:136)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:164)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:133)
	at org.janusgraph.core.JanusGraphFactory.open(JanusGraphFactory.java:80)
	at org.janusgraph.core.JanusGraphFactory$open.call(Unknown Source)
	... more
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.janusgraph.util.system.ConfigurationUtil.instantiate(ConfigurationUtil.java:58)
	... 54 more
Caused by: java.lang.IllegalArgumentException: The key 'root.index.elasticsearch.use-deprecated-multitype-index' cannot be false when existing index contains multiple types.
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:122)
	at org.janusgraph.diskstorage.es.ElasticSearchIndex.<init>(ElasticSearchIndex.java:235)
	... 59 more
```

For a migrated graph, [`useMultitypeIndex`](https://github.com/JanusGraph/janusgraph/blame/master/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java#L202) should be set to true.

--- a/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java
+++ b/janusgraph-es/src/main/java/org/janusgraph/diskstorage/es/ElasticSearchIndex.java
@@ -199,7 +199,6 @@ public class ElasticSearchIndex implements IndexProvider {
         useAllField = config.get(USE_ALL_FIELD);
         useExternalMappings = config.get(USE_EXTERNAL_MAPPINGS);
         createSleep = config.get(CREATE_SLEEP);
-        useMultitypeIndex = config.get(USE_DEPRECATED_MULTITYPE_INDEX);
         ingestPipelines = config.getSubset(ES_INGEST_PIPELINES);
         final ElasticSearchSetup.Connection c = interfaceConfiguration(config);
         client = c.getClient();
@@ -231,8 +230,14 @@ public class ElasticSearchIndex implements IndexProvider {
         } catch (final IOException e) {
             throw new PermanentBackendException(e.getMessage(), e);
         }
-        Preconditions.checkArgument(!useMultitypeIndex || !client.isAlias(indexName), "The key '" + USE_DEPRECATED_MULTITYPE_INDEX + "' cannot be true when existing index is split.");
-        Preconditions.checkArgument(useMultitypeIndex || !client.isIndex(indexName), "The key '" + USE_DEPRECATED_MULTITYPE_INDEX + "' cannot be false when existing index contains multiple types.");
+        if (!config.has(USE_DEPRECATED_MULTITYPE_INDEX) && client.isIndex(indexName)) {
+            // upgrade scenario where multitype index was the default behavior
+            useMultitypeIndex = true;
+        } else {
+            useMultitypeIndex = config.get(USE_DEPRECATED_MULTITYPE_INDEX);
+            Preconditions.checkArgument(!useMultitypeIndex || !client.isAlias(indexName), "The key '" + USE_DEPRECATED_MULTITYPE_INDEX + "' cannot be true when existing index is split.");
+            Preconditions.checkArgument(useMultitypeIndex || !client.isIndex(indexName), "The key '" + USE_DEPRECATED_MULTITYPE_INDEX + "' cannot be false when existing index contains multiple types.");
+        }
         indexSetting = new HashMap<>();
 
         ElasticSearchSetup.applySettingsFromJanusGraphConf(indexSetting, config, ES_CREATE_EXTRAS_NS);

--- a/janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchConfigTest.java
+++ b/janusgraph-es/src/test/java/org/janusgraph/diskstorage/es/ElasticSearchConfigTest.java
@@ -278,6 +278,23 @@ public class ElasticSearchConfigTest {
         idx.close();
     }
 
+    @Test
+    public void testMultiTypeUpgrade() throws InterruptedException, BackendException, IOException {
+        // create multi-type index
+        ModifiableConfiguration config = esr.setElasticsearchConfiguration(GraphDatabaseConfiguration.buildGraphConfiguration(), INDEX_NAME);
+        config.set(USE_DEPRECATED_MULTITYPE_INDEX, true, INDEX_NAME);
+        Configuration indexConfig = config.restrictTo(INDEX_NAME);
+        IndexProvider idx = open(indexConfig);
+        simpleWriteAndQuery(idx);
+        idx.close();
+
+        // should be able to open multi-type index if USE_DEPRECATED_MULTITYPE_INDEX is unset
+        config.remove(USE_DEPRECATED_MULTITYPE_INDEX, INDEX_NAME);
+        indexConfig = config.restrictTo(INDEX_NAME);
+        idx = open(indexConfig);
+        idx.close();
+    }
+
     private void simpleWriteAndQuery(IndexProvider idx) throws BackendException, InterruptedException {
         final Duration maxWrite = Duration.ofMillis(2000L);
         final String storeName = "jvmlocal_test_store";

